### ISSUE TITLE
einsum error for mixed subscripts on s390x

### ISSUE THREAD

**Issue by mbanck** [numpy/numpy#12689]
### Reproducing code example:

```python
import numpy as np
tensor = np.random.rand(10, 10, 10, 10)
np.einsum('ijij->', tensor)
```

### Error message:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3/dist-packages/numpy/core/einsumfunc.py", line 1346, in einsum
    return c_einsum(*operands, **kwargs)
ValueError: dimensions in operand 0 for collapsing index 'þ' don't match (19228356 != 10)
```

### Numpy/Python version information:

Debian unstable on s390x 

```
1.16.0rc2 3.7.2 (default, Jan  3 2019, 02:55:40) 
[GCC 8.2.0]
```

I could not reproduce this on Debian stable (numpy 1.12.1), but 1.15 seems to be affected as well.

So if I add `print(tensor[0][0][0])` before the  `np.einsum` it runs fine:

```
>>> import numpy as np
>>> tensor = np.random.rand(10, 10, 10, 10)
>>> print(tensor[0][0][0])
[0.56143868 0.22681118 0.47354192 0.27676805 0.09566204 0.63877983
 0.4243805  0.91690097 0.2695863  0.28861963]
>>> np.einsum('ijij->', tensor)
51.05293277361482
```

Also see https://github.com/psi4/psi4/issues/1465 where this bug originated for some more debugging output that might be helpful.

Note that on s390x, a `char` is `unsigned char` unless specified otherwise, contrary to x86 architectures.

**Comment by mbanck** [comment]
One more comment: I tried to example on code on armel (little-endian 32bit arm, which also has unsigned chars) and it segfaulted there:

```
Program received signal SIGSEGV, Segmentation fault.
0xb69c9e9c in get_combined_dims_view (labels=0xbeffda74 "ij\376\377\310\247C", iop=-1258924304, op=<optimized out>) at numpy/core/src/multiarray/einsum.c.src:2135
2135    numpy/core/src/multiarray/einsum.c.src: No such file or directory.
(gdb) bt full
#0  0xb69c9e9c in get_combined_dims_view (labels=0xbeffda74 "ij\376\377\310\247C", iop=-1258924304, op=<optimized out>) at numpy/core/src/multiarray/einsum.c.src:2135
        i = -1231350260
        label = 254
        new_strides = {8080, 800, 0, 0, -1228307968, 0, 3997012, -1228357744, 767293889, 1735752, 2, -1228357696, 1645376, 4434692, 1721428, 0, -1254964384, -1254964076, 4520024, 
          4520024, 1196628, 0, 4517992, -1228721336, 1189788, 0, -1228727728, 1651712, -1228307968, -1228684220, 3997012, -1228357744}
        new_dims = {10, 10, 0, 0, 1626092, 0, 3923600, -1255724560, -1255740552, -1255724560, -1228505040, 4434692, 0, 21, 4012324, 4436364, 1648256, 0, -1228357744, 2, 5, 
          -1228684220, -1228417574, 4434692, 4436364, 4012324, -1228684544, 4520024, 1219524, -1255724560, -1255740552, -1228417584}
        ndim = 4
        icombine = 2
        ret = 0x0
        idim = <optimized out>
        combineoffset = -1
        icombinemap = {0, 1, -1231350260, 12, -1090527424, -1090527420, 1735752, 0, 1128304, 3919576, -1090527424, -1090527420, 0, 0, 4, 1, 3833176, -1229092192, -1228731352, 4012324, 
          -1255738952, 0, 2, -1090527452, -1254699360, -1228693264, 4, 0, 79, 0, 750306304, 0}
        new_strides = <optimized out>
        new_dims = <optimized out>
        idim = <optimized out>
        ndim = <optimized out>
        icombine = <optimized out>
        combineoffset = <optimized out>
        icombinemap = <optimized out>
        ret = <optimized out>
        label = <optimized out>
        i = <optimized out>
#1  PyArray_EinsteinSum (subscripts=<optimized out>, subscripts@entry=0x3bced8 <__stack_chk_guard> "", nop=nop@entry=1, op_in=op_in@entry=0xbeffef84, 
    dtype=0x3bced8 <__stack_chk_guard>, order=NPY_KEEPORDER, casting=NPY_SAFE_CASTING, out=out@entry=0x0) at numpy/core/src/multiarray/einsum.c.src:2672
        labels = <optimized out>
        combine = <optimized out>
        ndim = <optimized out>
        iop = -1258924304
        label = <optimized out>
        min_label = 0
        max_label = -1090529416
        label_counts = '\000' <repeats 105 times>, "\002\002", '\000' <repeats 20 times>
        op_labels = {"ij\376\377\310\247C\000\000\000\000\000\064\304\303\266(\371\307\266P\021\303\266|\261C\000d\321\277\266", 
          "\377\377\377\377(|\312\266@\304\303\266\377\377\377\377h\360D\000\004\253C\000\360\332\377\276\200\241\303\266", 
          "\334\003\303\266\310\247C\000\000\000\000\000d,\304\266H|\032\000\331c\312v\\\344;\000X\370D", 
          "\001\000\000\000\000\000\000\000\070\304\303\266\004\253C\000TD\032\000\000\000\000\000\000\303\303\266\064\304\303\266", 
          "X\370D\000\000\303\303\266x\r\030\000X\370D\000X\370D\000@\304\303\266\360\065\022\000\000,'\265", 
          "|\333\377\276\001\000\000\000\334\272\305\266\220\336;\000(\a\303\266\260*'\265\000\000\000\000h\371\303\266", 
          "\000]\037\335\200\024\312\266<\345;\000\246\003\000\000\060\360\030\000(|\312\266\200\024\312\266\000\000\000", 
          "\220\367\303\266P\347\066\265X\n\257\252P*'\265\000T\f \\\344;\000\344\314\030\000\370\202\304\266", 
          "\000\000\000\000\001\000\000\000x\343;\000\000\000\000\000(\a\303\266\210\005\031\000\360\333\377\276\000\257\303\266", 
          "T\216\302\266\345\000\000\000\250\201\304\266X\370D\000\060\216\302\266\000\000\000\000T\375<\000\320\256\303\266", 
          "y\327Z\236H|\032\000\001\000\000\000\000\257\303\266@\033\031\000$9=\000\030Q\022\000\000\000\000", 
          "\001\000\000\000\230q\307\266\f\334\377\276\377\377\377\377\210\334\377\276\000\000\000\000\000\000\000\000\003\000\000", 
          "X\334\377\276\360\251\303\266\304\002\303\266\000\000\000\000\367S\343\245\000\000\000\000\240\002\303\266\000\000\000", 
          "T\375<\000P\032\303\266M\211w]H|\032\000\001\000\000\000\360\251\303\266@\033\031\000\001\000\000", 
          "\002\000\000\000T\375<\000\260\254L\265\004\253C\000x\r\030\000$9=\000\260\254L\265\220\336;", 
          "dT\034\000X\005\304\266\000\000\000\000\000\000\000\000\367S\343\245\000\000\000\000\240\002\303\266T\375<", 
          "T\375<\000P\032\303\266\000\024\f\020$9=\000\200\304\302\266\000\000\000\020\220\315\030\000 \241\303\266", 
          "\304\243&\265x\022D\000\324>C\000\000\000\000\000Lw\032\000 \241\303\266\\\344;\000\220\336;", 
          "\320O\307\266\000\000\000\000\004\253C\000\310GC\000$9=\000\004\000\000\000\314j\020\000T\375<", 
          "\020\363\376\266\224\335\377\276\370O\034\000$9=\000\200\304\302\266X\370D\000T\a\031\000\001\000\000", 
          "\230\335\377\276X\005\304\266\304\305\302\266\000\000\000\000\310\305\302\266\000\000\000\000\000\000\000\000\230\335\377\276", 
          "\324\026\022\000\000\000\000\000P\032\303\266(\253C\000\000\000\000\000\310\305\302\266 \325\307\266\004\253C", 
          "&\000\000\000$9=\000\200\304\302\266X\370D\000`\224\022\000\220\336;\000\037\000\000\000\370\324\307\266", 
          "\220\336;\000\310\247C\000\006\000\000\000\000\000\000\000\177\000\000\000dR8\000\000\000\000\000\334\335\377\276", 
          "\274\336\377\276`\273\070\000\230\060\026\000\377\377\377\377df\034\000\240@\310\266\b\362\303\266\020\347\066\265", 
          "X\005\304\266\304\305\302\266P\201-\265\230\217&\265H|\032\000\060\211$\265-\331\304\254\\\344;", 
          "0\211$\265\020\347\066\265f\273\070\000\340\001\000\000Q\000\000\000\070\000\000\000\070\000\000\000\000g6\265", 
          "<\253C\000\330\316;\000<\337\377\276\070\337\377\276\230,\026\000\364d7\000x\353&\265\023\000\000", 
          "\000g6\265y}:\000gfff\364d7\000\377\377\377\377\000\000\000\000\000\000\000\000U\000\000", 
          "\240\373\023\000y}:\000y}:\000gfff\230U\026\000pz\312\266\000\000\000\000\300\336\377\276", 
          "\001\000\000\000\000\000\000\000S\000\000\000\334FL\265T{\034\000\001\000\000\000\000\000\000\000\f\000\000", 
          "\001\000\000\000\177\000\000\000\205\000\000\000\200\374\263\266H\374\263\266\000\340\263\266\001\000\000\000\v\000\000"}
        output_labels = "`\224\022\000\220\336;\000\034z\032\000\370\324\307\266\220\336;\000\310\247C\000\000\000\000\000\264\305\302\266"
        iter_labels = <optimized out>
        idim = <optimized out>
        ndim_output = 0
        ndim_broadcast = <optimized out>
        ndim_iter = <optimized out>
        op = {0x0, 0x0, 0xb4f66208, 0xb4f6b210, 0x1c85fc <_PyObject_FastCallDict+88>, 0x0, 0xb5268fa8, 0xb5268f58, 0xb5268fa8, 0x3d308c <PySuper_Type>, 0x3d0cd8 <PyDict_Type>, 
          0x1a7058 <dict_dealloc>, 0xb5270fd0, 0x3bde90 <_Py_NoneStruct>, 0x179b84 <CFLOAT_sign+292>, 0x0, 0x3d0cd8 <PyDict_Type>, 0xbeffd7fc, 0x1, 0x44f858, 0x194, 0xb4f66208, 
          0xb4f6639c, 0x0, 0x0, 0x0, 0xb4f66278, 0xb4f66244, 0xb4f662a4, 0xb4f6b1b0, 0x0, 0x43ab3c <_Py_AllocatedBlocks.lto_priv.1425>}
        ret = 0x0
        op_dtypes_array = {0xa, 0x0, 0xb5272ba0, 0x0, 0xb5373750, 0xb6c36fc0, 0xb5268f30, 0xb4f66208, 0x24c, 0x1, 0x0, 0x41, 0x7, 0x0, 0x0, 0xb6e5a7d4, 0x5c, 0xb6e5a258, 0x6e, 
          0x44e128, 0x41, 0x260, 0x9, 0x4a, 0x1, 0x247, 0x0, 0x77, 0x0, 0x7c, 0x3d36cc <PyType_Type>, 0x1a7058 <dict_dealloc>}
        op_dtypes = <optimized out>
        op_axes_arrays = {{0, 1, -1231350260, 12, -1090527424, -1090527420, 1735752, 0, 1128304, 3919576, -1090527424, -1090527420, 0, 0, 4, 1, 3833176, -1229092192, -1228731352, 
            4012324, -1255738952, 0, 2, -1090527452, -1254699360, -1228693264, 4, 0, 79, 0, 750306304, 0}, {6105552, 4, -1090527284, 6105552, 4, -1090527284, 0, 0, -1231557396, 1, 4, 
            0, -1090527420, -1090527416, 0, 1, 3919576, 5006840, 1, -1231163720, 0, 0, 1, 4520024, 0, -1090527364, -1228298160, 4520024, -1090527276, -1228496848, -1090527276, 
            4434692}, {-1228298160, 4012324, -1241730920, -1254699600, 0, -1242996688, -1230255840, 4, 0, -1229718456, -1231240204, 0, 4, -1254672400, -1229718456, -1229725696, 0, 0, 
            -1231474916, -1, 4520024, 4012324, 0, -1228692064, 0, -1228295532, -1241730920, 1639816, -1230255840, 8, 0, -1229718400}, {-1231240204, 0, 8, -1254672480, -1229718400, 
            -1229725696, 0, 0, -1231474916, 4434692, -1253292240, 4012324, 1200408, 4434692, -1230255840, 8, 0, -1229718400, -1231240204, 0, 8, -1254672400, -1229718400, -1229725696, 
            0, 0, -1231474916, -1, 2, -1228705144, -1090527056, -1243026912}, {-1228706060, -1228692064, 3923600, 0, 3919576, -1229723548, 0, 0, 0, 4520024, -1230255840, 4, 0, 
            -1229718736, 0, 0, 4, -1254672400, -1229718736, -1229725696, 8, 0, 750306304, 56, -1254672360, -1229718456, 0, 0, 6105552, 3919576, 8, 1}, {0, 56, 7, 0, 16, -1226463276, 
            91, -1226464680, 110, 4513800, 2, 32, 0, 2, 0, 7, 0, 119, 0, 124, 5006848, 5006840, -1226463276, 0, -1226465280, 8, 0, 0, 0, 0, -1227362148, 5006856}, {-1229559328, 8, 
            -1230255840, 8, -1231729800, 8, -1231240204, 0, -1229652952, -1229659028, -1230754428, -1255964944, 750306304, 0, -1229718400, 3919576, 0, -1255964944, -1090526628, 
            -1255964944, 1, 1, -1231473176, 0, 0, 0, -1254672400, 0, -1255964936, -1090526628, 5006856, -1229718400}, {-1230751344, 0, 0, 0, 0, 3919576, 0, 0, -1229725696, 3919576, 
            -1230255840, 8, -1230291708, -1229718400, -1231240204, 0, 8, -1254672480, -1229718400, -1229725696, 0, 0, -1231474916, 0, -1229718400, 0, 0, 1073741824, -1229718456, 2, 1, 
            6201932}, {3919576, -1229723548, 0, 0, 0, -1090525488, 0, 3919576, -1241693936, 2, 3, -1090525488, -1230293128, 3919576, -1229833268, -1229725696, 750306304, -1090526396, 
            -1090525492, 5010104, -1229725696, 3919576, 6653832, 5006840, -1230522508, 0, 750306304, 1, 0, -1255964944, -1090525332, -1229136804}, {-1229723548, -1229725696, 
            -1230467428, 0, 0, 5006856, 1, 3919576, -1230428952, 0, 0, 0, 0, 1, 750306304, 0, 0, 0, 2, 2, -1230441676, 0, 0, 0, 0, -1090526296, 1, 0, 5006856, -2147483648, 
            -1230517236, 1}, {0, 0, -1090526284, -1090526124, -1090525332, -1090526284, 1, -1230517352, 16777216, -1090525332, -1230270668, 0, 0, 1, 5006856, 3919576, -1230428952, 1, 
            0, 0, 0, 1, 750306304, 0, 0, -1230291708, 1, 1, -1230441676, 0, 0, 0}, {0, -1090526136, -1090526124, 0, 0, 0, 127, 3691108, 2, 3919576, -1229956576, -1090526068, 8192, 
            -1090525940, -1090526084, 0, 40, -1090526008, -1228763600, 0, 4878696, 1, 8192, 521, 0, 2, 3, 0, 0, -1230517352, 0, 0}, {-1229136816, 0, -1229136816, 0, 62, -1258913592, 
            1074921472, 1259601920, -1254681448, 13, -1258913592, 3824173, 1717986919, 3630324, -1, 0, 0, 85, 1309600, 3824173, 3824173, 1717986919, 1463704, 127, 40, -1090525848, 
            4434692, 0, -995973801, 2112131729, 1721226987, -588140404}, {0, -1258913568, 1, 127, 141, 62, -1229718400, -1229718400, 0, 7, 0, -1254681448, 0, 3927276, 3924652, 
            4011724, 11268, 0, 1128304, 818938375, -1090525856, 0, -1241630416, 0, -1229652952, 1527012, 3824132, -1229811864, -1241630416, -1254672376, 0, 4011724}, {-1229460744, 0, 
            2, -1229519228, 2, -4, 0, 0, 0, -1229994492, 4011724, -1254681448, 0, -1241630416, 1126332, 0, -1229652952, -1241630416, -1230238288, 0, -1255964944, 0, -1230280672, 4, 
            -1228749696, -1229136816, 0, -1229801224, 3919576, -1090525360, 3923600, 4878696}, {6, 0, 0, 3919576, -1241630496, -1090525580, -1090525356, 3717984, 16, -1229652952, 8, 
            0, -1255964936, 8, 0, 3919576, -1255964936, 0, -1231661080, -1229718400, -1255964944, 8, -1230749892, 61, -1231661108, 2, 750306304, -1254672400, -1255964944, 0, 
            -1090525516, -1090525232}, {-1231715288, -1254672400, -1255964944, 0, -1230748564, -1090525096, 5010104, 0, -1230437164, -1254795344, 0, 2, 3919576, -1090525504, 
            -1090525492, 0, 16, -1229652952, 8, 0, -1255964920, 8, 0, 3919576, -1255964920, 0, -1231661080, -1229718400, -1255964928, 8, -1230749892, 61}, {-1231661108, 1, 750306304, 
            -1254672480, -1255964928, 0, -1090525356, -1090525072, -1231715288, -1254672480, -1255964928, 0, -1230748564, -1090524940, 4878696, 0, -1230437164, -1229136816, 0, 
            -1254795344, 3919576, -1090525344, -1090525332, 0, -1229136816, 0, 4878696, -1229136816, 0, 0, -1254672400, 0}, {-1253121280, -1242733072, 0, -1090525172, 1185492, 
            -1242733072, 750306304, 0, -1254672880, 32, -1090525240, 0, 0, -1254672880, -1254672880, -1230950864, 750306304, 3919576, -1254672880, -1229725696, 11, 0, 0, -1254672880, 
            -1254672880, 256, -1230929632, 256, 0, -1242758256, -1255964928, -1}, {4520024, 4434692, -1090525120, 3919576, -1228706060, 4434692, -1253121284, 4012324, 1735752, 0, 
            3925084, 4520024, -1251351056, -1242051056, -1251351056, 4434692, 1721428, -1254963260, -1253121608, -1253121300, 4520024, 4520024, 750306304, 0, 5010104, -1090525044, 
            -1230437936, 5010104, -1090525044, -1255964944, 0, 5010104}, {1872132, 4, 16, -1229652952, 3919576, -1090524960, -1255964928, -1228967712, -1251351376, 3919576, 750306304, 
            -1228967712, 3919576, -1255964928, -1255964928, 1835992, -1230889644, -1228966832, 1, 0, -1229997748, -1230404904, -1230404924, -1229725696, 3919576, -1255964928, 1, 0, 
            -1229997748, -1230309832, -1230310124, 0}, {0, -1998362383, 1055193269, 4012324, 750306304, 4434692, 0, -1228400160, 4012324, -1228966832, 1648816, 0, -1230350432, 
            3923588, 750306304, -1255964928, -1251351056, -1228294542, 4434692, -1229659640, -1251351056, 42, 1219100, -1228842272, 2, -1228294576, 3923600, 4433864, 0, -1251351068, 
            -1228803536, -1229288896}, {-1255964928, -1251351056, -1, -1228774336, -1251351060, -1, 3923600, -1243015952, 1871916, -1243026912, -1241629120, -1251351056, -1251351060, 
            -1090524664, -1090524656, -1243026912, -1229081516, 1852404851, -1243651112, -1229136816, 1, -1090524692, 750306304, -1254672880, -1251351068, 4520024, 3923600, 
            -1251351060, -1231715288, -1251351376, 0, 3672268}, {-1090523804, 0, 1111056, -1228825504, -1228707540, 0, -1229145736, -1228825504, -1228505040, 3672272, 0, 1734476, 
            -618135140, 0, 3672269, -1228707576, 1729600, 4012324, -1090524600, 0, 0, -1228678116, 4648938, 4434692, 664, 4012324, 1214328, 0, 3997012, 4648272, 3923600, 4433864}, {0, 
            -1228678116, -1228606848, -1228285936, -1228678116, -1228678112, -1, -1228773280, -1228678116, -1, -1090524416, -1228357696, -1228307932, -1228825504, -1228453648, 
            -1228678116, -1228307968, 0, 3997012, -1228357744, 767293889, 1735752, 2, -1228357696, 1645376, 4434692, 1721428, 0, -1228678424, -1228678116, 4520024, 4520024}, {1196628, 
            0, 4517992, -1228721336, 1189788, 0, -1228727728, 1651712, -1228307968, -1228684220, 3997012, -1228357744, 269227008, 4012324, -1228308032, -1228325472, 1626092, 0, 
            3923600, -1228825504, -1228773280, -1228825504, -1228505040, 4434692, 0, 7, 4012324, 4436364, 1648256, 0, -1228357744, 2}, {5, -1228684220, -1228417574, 4434692, 4436364, 
            4012324, -1228684544, 4520024, 1219524, -1228825504, -1228773280, -1228417584, 3923600, 4433864, 0, -1228684236, -1228408536, -1228730032, 4436348, -1228943004, -1, 
            -1228243928, -1228684224, -1, 4517992, 4434692, -1090524128, -1228693120, -1228733476, 4433864, 0, -1228657564}, {1735752, 1992975321, 3925084, 4520024, 1, 0, -1228684232, 
            4434692, 1721428, 0, -1228684544, -1228684236, 4520024, -1228684544, 1576312, 4520024, 4520024, -1228684224, 1193456, -1229094528, -1090523988, 1, -1228555556, 3923600, 
            -1228732632, -1229123648, 0, -1228670616, -585147136, -1228270464, 3925308, 935}, {1634352, -1228243928, -1228270464, 0, -1228671088, -1229138672, -1431369128, 
            -1229123840, 537678848, 3925084, 1625316, -1228635400, 0, 1, 3924856, 0, -1228732632, 1639816, -1090523872, -1228689664, -1228763564, 229, -1228635736, 4520024, 
            -1228763600, 0, 3997012, -1228689712, -1638213767, 1735752, 1, -1228689664}, {1645376, 4012324, 1200408, 0, 1, -1228443240, -1090523844, -1, -1090523720, 0, -1229652952, 
            -1229659028, -1230754428, -1229725696, -1229652952, -1090522772, 3919576, -1229725696, -1231385920, 0, 3997012, 4, 4878832, 4, 4878832, 4, -1231484664, 128, 0, 4, 
            -1229092656, -1090522764}, {-1229725696, -1090522796, 750306304, -1090522764, -1229959827, -1090523604, -1229959829, -1229652952, -1231387616, -1090522788, 2147483647, 
            -1090522772, 1105176, -1228727728, 269227008, 4012324, 0, 3672268, -1090522812, -1, 1111056, 4461176, 4406996, 0, 1734476, -1228693216, -1228505040, 3672272, 0, 2, 
            3919576, -1090523540}, {750306304, 64, -1229082352, -1229718400, 0, -1229725696, -1229718400, 2, -1231472664, -1229725696, 3919576, 0, 0, -1090523540, -1229082352, 0, 
            -1231457580, -1090523544, -1090523540, -1090523552, 0, 0, -1228727728, 4434728, 0, -1228749368, 0, -1229082352, 0, 0, -1228749696, 4520024}}
        op_axes = {0xb6e5a7d4, 0xb6fef7f8, 0xb6e5a000, 0x24c, 0x48, 0x1f8, 0xb532c360, 0x40, 0xb6d7f09c <malloc+164>, 0x24b, 0x24c, 0x0, 0xb54e2870, 0x1888dc <_PyMem_RawMalloc>, 
          0x189b68 <PyObject_Malloc+216>, 0xb532c4a0, 0x0, 0xb532c4a0, 0x0, 0x3808cc, 0xbeffdc34, 0x0, 0x10f410 <_array_fromobject>, 0xb52729f0, 0xb52685e4, 0x0, 0xb4f67270, 
          0xb52729f0, 0xb6c68030, 0x3808d0, 0x0, 0xb4f67270}
        iter_flags = <optimized out>
        op_flags = {8080, 800, 0, 0, 3066659328, 0, 3997012, 3066609552, 767293889, 1735752, 2, 3066609600, 1645376, 4434692, 1721428, 0, 3040002912, 3040003220, 4520024, 4520024, 
          1196628, 0, 4517992, 3066245960, 1189788, 0, 3066239568, 1651712, 3066659328, 3066283076, 3997012, 3066609552}
        iter = <optimized out>
        sop = <optimized out>
        fixed_strides = {10, 10, 0, 0, 1626092, 0, 3923600, -1255724560, -1255740552, -1255724560, -1228505040, 4434692, 0, 21, 4012324, 4436364, 1648256, 0, -1228357744, 2, 5, 
          -1228684220, -1228417574, 4434692, 4436364, 4012324, -1228684544, 4520024, 1219524, -1255724560, -1255740552, -1228417584}

**Comment by seberg** [comment]
@mbanck the code there actually mentions the sign of char, but on the second similar occurrence the cast is missing, can you try if this fixes it?:
```diff
diff --git a/numpy/core/src/multiarray/einsum.c.src b/numpy/core/src/multiarray/einsum.c.src
index 1765982a0..24e2794f5 100644
--- a/numpy/core/src/multiarray/einsum.c.src
+++ b/numpy/core/src/multiarray/einsum.c.src
@@ -2115,7 +2115,7 @@ get_combined_dims_view(PyArrayObject *op, int iop, char *labels)
         /* If this label says to merge axes, get the actual label */
         if (label < 0) {
             combineoffset = label;
-            label = labels[idim+label];
+            label = (signed char)labels[idim+label];
         }
         else {
             combineoffset = 0;
```

**Comment by seberg** [comment]
Or actually, maybe the exact opposite, making it `signed char` there, would have to take a bit more time to understand the code.

**Comment by mbanck** [comment]
> Or actually, maybe the exact opposite, making it `signed char` there, would have to take a bit more time to understand the code.

Yeah, I had to remove the `(signed char)` above, then it seems to run fine on a quick check; also the psi4 testcase that was failing earlier now passes.

**Comment by charris** [comment]
I suspect it just needs to be consistent when parsing the string, the actual values probably don't matter that much.

**Comment by dgasmith** [comment]
Original issue here: https://github.com/psi4/psi4/issues/1465

@mbank did you want to make this change?

**Comment by mbanck** [comment]
I was diffing through the `einsum.c.src` changes between 1.12 and 1.16 in order to find the difference as I said:

> I could not reproduce this on Debian stable (numpy 1.12.1)

But I have now successfully reproduced it on armel on 1.12.1:

```
>>> import sys, numpy; print(numpy.__version__, sys.version) 
1.12.1 3.5.3 (default, Sep 27 2018, 17:25:39) 
[GCC 6.3.0 20170516]
>>> import numpy as np
>>> tensor = np.random.rand(10, 10, 10, 10)
>>> np.einsum('ijij->', tensor)
Segmentation fault
```

So this has been broken for a longer time. I don't feel qualified to comment on the correct change though.

**Comment by charris** [comment]
Given the check for negative, it seems the negative values are probably used as flags

**Comment by charris** [comment]
I can compile with the `-funsigned-char` flag in fedora and all the tests pass, but the example here segfaults. It's iffy to draw conclusions from that -- many things could be going wrong -- but the proposed fix doesn't do the job.

**Comment by charris** [comment]
Oh, and the negative values are used as a flag, so we need to keep that.

**Comment by charris** [comment]
The easiest fix might be to simply convert the labels to `signed char` along with all the other `char` related declarations.

**Comment by mbanck** [comment]
I'm happy to test a patch - I think the tests should be expanded as well though with something like the testcase I posted initially

**Comment by jlatone** [comment]
> > Or actually, maybe the exact opposite, making it `signed char` there, would have to take a bit more time to understand the code.
> 
> Yeah, I had to remove the `(signed char)` above, then it seems to run fine on a quick check; also the psi4 testcase that was failing earlier now passes.

I had the same issue on POWER9. Thanks for this helpful thread and the (for now) fix! --joe

**Comment by charris** [comment]
Just to add a bit of history, the original code assumed `char` was signed, which worked fine until an architecture showed up with unsigned `char` -- the C89 standard leaves it up to the implementation. Whereupon there were a lot of fixups. I don't know that the current problem is related to that, but it smells related.

**Comment by mattip** [comment]
Fixed in #14693
